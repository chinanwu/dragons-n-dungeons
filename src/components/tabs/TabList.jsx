import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import getThemeClassName from '../../functions/getThemeClassName';
import TabButton from './TabButton.jsx';

import './TabList.less';

export const TabList = ({
  list,
  activeIndex,
  header,
  footer,
  theme,
  onClick,
  onKeyDown,
}) => {
  let tabButtons = [];

  if (list !== undefined && list.length !== 0) {
    tabButtons = list.map(({ name, disabled }, index) => {
      return (
        <TabButton
          id={`tabButton__${index}`}
          label={name}
          index={index}
          isActive={index === activeIndex}
          isDisabled={disabled}
          onClick={onClick}
          onKeyDown={onKeyDown}
          key={`TabButton__${index}`}
        />
      );
    });
  }

  return (
    <div
      className={getThemeClassName('TabList', theme)}
      role="tablist"
      aria-label="Tabs"
    >
      <div className="TabList--left">
        {header}
        {tabButtons}
      </div>
      <div className="TabFooter">{footer}</div>
    </div>
  );
};

TabList.propTypes = {
  list: PropTypes.array,
  activeIndex: PropTypes.number,
  header: PropTypes.any,
  footer: PropTypes.any,
  theme: PropTypes.string,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
};

export const mapStateToProps = ({ theme }) => ({
  theme: theme.theme,
});

export default connect(mapStateToProps)(TabList);
