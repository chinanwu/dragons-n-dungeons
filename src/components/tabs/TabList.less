@import '../styling/colours.less';

.TabList {
  display: flex;
  justify-content: space-between;
  padding-top: 10px;
}

.TabList--light {
  border-bottom: @tablist-border-width @tablist-border-style
    @tablist-border-colour;
  background-color: @tablist-background-colour;
}

.TabList--dark {
  background-color: black;
}

.TabList--left {
  display: flex;
}

.TabFooter {
  padding-right: 20px;
}
