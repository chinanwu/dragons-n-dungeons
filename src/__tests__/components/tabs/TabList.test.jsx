import { shallow } from 'enzyme';
import React from 'react';
import { TabList, mapStateToProps } from '../../../components/tabs/TabList.jsx';

jest.unmock('../../../components/tabs/TabList.jsx');
jest.unmock('../../../functions/getThemeClassName');

describe('TabList component', () => {
  describe('rendering', () => {
    it('renders TabList', () => {
      const wrapper = shallow(<TabList />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('renders TabList with empty list', () => {
      const wrapper = shallow(<TabList list={[]} />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('renders TabList with list', () => {
      const wrapper = shallow(
        <TabList list={[{ name: 'test-tab', disabled: false }]} />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    it('returned mapped properties', () => {
      const theme = 'light';
      const state = { theme };
      expect(mapStateToProps({ theme: state })).toEqual({ theme: theme });
    });
  });
});
