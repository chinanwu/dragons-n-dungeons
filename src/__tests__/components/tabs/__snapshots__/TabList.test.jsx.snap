// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabList component rendering renders TabList 1`] = `
<div
  aria-label="Tabs"
  className="TabList"
  role="tablist"
>
  <div
    className="TabList--left"
  />
  <div
    className="TabFooter"
  />
</div>
`;

exports[`TabList component rendering renders TabList with empty list 1`] = `
<div
  aria-label="Tabs"
  className="TabList"
  role="tablist"
>
  <div
    className="TabList--left"
  />
  <div
    className="TabFooter"
  />
</div>
`;

exports[`TabList component rendering renders TabList with list 1`] = `
<div
  aria-label="Tabs"
  className="TabList"
  role="tablist"
>
  <div
    className="TabList--left"
  >
    <Memo(Connect(TabButton))
      id="tabButton__0"
      index={0}
      isActive={false}
      isDisabled={false}
      label="test-tab"
    />
  </div>
  <div
    className="TabFooter"
  />
</div>
`;
